### PR TITLE
[DRAFT] BB_LOG: A better logging API with arbitrary structured JSON to build stuff with

### PR DESCRIPTION
**Pull request by scriptis**
## About The Pull Request

**This is a draft.** Not everything is implemented, and this is a call for comments on this sort of approach.

### Gimme feedback. Talk about this and whether it's a good idea.

This is the start of a (rather large) refactor that replaces server logging facilities with a single macro, `BB_LOG`, that accepts a format string and arbitrary context arguments. The format string is rendered with the provided context, resulting in text logs. Downstream consumers can either use the plaintext formatted render, or use the JSON value directly. This, for example, would allow for a tgui-based log viewer with actually intelligent filtering or jumping to arbitrary players/turfs/weapons/spent projectiles (which is why I started this draft).

This PR starts with adding the `BB_LOG` macro *adjacent* to all combat logging. For example,

```
// present:

log_combat(user, target, "attacked", "[item.name]", "(Combat mode: [user.combat_mode ? "On" : "Off"])")
```

```
// would be replaced with, but is currently just followed by:

BB_LOG( \
	BB_COMBAT, \
	"{user} stabbed {target} in the eyes with {with}", \
	user = user, \
	target = target, \
	with = item, \
)

// combat mode is associated with `/mob/living` via `/mob/living/bb_snapshot()`
// and is automatically injected into the raw JSON output, along with a number of
// other variables an admin might care about (health, stamina, species, name, realname, whatever)
//
// {logdir}/bb.context.log receives the raw JSON payload
// {logdir}/bb.rendered.log receives just the rendered text similar to the current log output
```

Sample outputs (only `COMBAT` and `EMOTE` are currently implemented):

[Raw JSON](https://pastebin.com/raw/RapNza7g)
[Rendered text](https://pastebin.com/raw/i1h7Vjff)

At present, the objective is to eventually override the current `*.log`s with individual forked, rendered outputs, and have all of the raw JSON in `bb.log`. This preserves the current logging infrastructure and deduplicates work.

The macro is defined in `code\__DEFINES\bb_defines.dm`. The actual work is done in `code\modules\logging\bb.dm`.

gimme feedback
## Why It's Good For The Game

Makes logging much easier. Allows log consumers to get more information than they would get out of just plaintext logs. Gets me halfway to a better admin log viewer than [this abomination](https://i.imgur.com/7CRfl1S.png).
## Changelog
:cl:
refactor: the entire logging infrastructure, eventually
/:cl:

### CODE REVIEW

**Comment by scriptis** [comment]
tl;dr: I want to right-click on the floor next to the supermatter engine and click "BB: Show all events near here" and see who turned off the supermatter vents in under 4 seconds

**Comment by Mothblocks** [comment]
https://github.com/tgstation/tgstation/pull/73604/

???

**Comment by Mothblocks** [comment]
I cannot help but notice you made 3,000 additions to logging code without realizing your exact same project has been done better and to spec by someone else weeks ago. Maintainers should've told you this if you said anything about this while working on it

**Comment by Mothblocks** [comment]
Reasons this is worse than the other PR which we will be merging soon assuming no other complications:

1. It has a version field. Without one it is impossible to build smart tooling
2. It will be split by exact actions. It looks like your code is completely ambiguous as it's all under "combat" and would still require English parsing, which is never acceptable
3. It doesn't use a macro which means it looks like fairly normal code, rather than littered with backslashes

**Comment by Mothblocks** [comment]
> Gets me halfway to a better admin log viewer than [this abomination](https://i.imgur.com/7CRfl1S.png).

I also really don't like this view.

We need better external tools, absolutely. But our existing tools should be just as good and usable. They are the ones that *all* admins, from today and in several years from now will be using. They need to be acceptable.

**Comment by scriptis** [comment]
Harsh, jeeze. This was just an afternoon's work, don't stress on it, but

- This is just combat because it's a draft, that's the point. This was the "hey can I get some feedback on this idea" thread.
- The goal is to actually have contextual objects (i.e., be able to hover over a button in the log viewer to jump to an attacker, defender, the weapon they were using, whatever, and be able to hover to see their health/stam/&c), which is not accomplished by Zephyr's PR.
- I wanted to replace the logging procs with something actually human readable (escaped strings).
- Zephyr's PR is just the existing logging infrastructure except it's a JSON object with a timestamp, the original log message, and an arbitrary list?

**Comment by scriptis** [comment]
Hey, I'm cool with moving this work over to Zephyr's work. I'd appreciate it if you didn't just come out swinging, though. I don't follow the PR tracker, and most of this PR is just the same macro copy-pasted to everything. There's only two files I care about getting feedback on, those are the two listed in the OP.

**Comment by Mothblocks** [comment]
> Zephyr's PR is just the existing logging infrastructure except it's a JSON object with a timestamp, the original log message, and an arbitrary list?

So that we can get the system in, yes. The end-goal is everything on a JSON log, like what you are building except without maintaining two sources of truth.

> The goal is to actually have contextual objects (i.e., be able to hover over a button in the log viewer to jump to an attacker, defender, the weapon they were using, whatever, and be able to hover to see their health/stam/&c), which is not accomplished by Zephyr's PR.

It very easily could/should be! As I said, the goal as I understand it is to make everything a JSON log.

**Comment by Mothblocks** [comment]
I also don't think I was being rude I was just very confused on if this was a project that maintainers have been talking about in Discord that I had no idea about

**Comment by scriptis** [comment]
Yeah, well, I wasn't aware that the other PR existed and you came out calling this worse. You provided a list of reasons why it's worse. A list!

Eh, don't stress on it, shit happens.

I'm just going to put this off until Zephyr's stuff is merged and stable, then build from there. It sucks, but yeah, this seems like it was wasted effort.

**Comment by optimumtact** [comment]
> I also don't think I was being rude

>I cannot help but notice you made 3,000 additions to logging code without realizing your exact same project has been **done better** and to spec by someone else weeks ago

**Comment by Mothblocks** [comment]
I see, I apologize for how that came off

**Comment by scriptis** [comment]
Yeah, it'll be fine. I've had a rough couple of weeks myself, sometimes stuff just gets to you. I've gotten in touch with Zephyr to see about salvaging some of this--I'd at least like to get the template strings in, since those would be key to a decent tgui log viewer.

File: code/modules/logging/bb.dm [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+			if(!isnull(value))
+				if(!istext(key) && !isnum(key))
+					// We can't store non-numeric/non-string keys in JSON maps.
+					// Skip this one.
+					continue
+
+				.[key] = _bb_jsonify(data[key])
+			else if(!isnull(key))
+				. += list(_bb_jsonify(key))
```
Comment (LemonInTheDark) [inline code/modules/logging/bb.dm:25]: This isn't actually safe as an assumption.
byond doesn't distinguish between an array and a map, they are the same thing and can be intermixed.
all this try/catch block will catch is out of bouds accesses, but if we have a list like this
`list(2, "jerry I'm a flat list", "bill I know you told me yesterday")`
This code would runtime on line 23 (due to an index out of bounds exception)
